h/services/bulk_api/lms_stats.py

```diff
@@ -17,65 +17,83 @@ class AssignmentStats:
-                (func.array_length(Annotation.references, 1) != None, "reply"),
-                # Not anchored, page note
-                (
-                    func.jsonb_array_length(Annotation.target_selectors) == 0,
-                    "page_note",
-                ),
-                # No annotation text, highlight
-                (func.length(Annotation.text) == 0, "highlight"),
-                # Anything else, an annotation
-                else_="annotation",
-            ).label("type"),
-        ).join(Annotation)
-
-    def assignment_stats(
-        self, groups: list[str], assignment_id: dict
-    ) -> list[AssignmentStats]:
-        """
-        Get basic stats per user for an LMS assignment.
-
-        :param groups: List of "authority_provided_id" to filter groups by.
-        :param assignment_id: ID of the assignment we are generating the stats for.
-        """
-
-        annos_query = (
-            self._annotation_type_select()
+    def _annotation_query(self, groups: list[str], assignment_id: str | None = None):
+        query = (
+            select(
+                AnnotationSlim,
+                AnnotationMetadata.data,
+                case(
+                    # It has parents, it's a reply
+                    (func.array_length(Annotation.references, 1) != None, "reply"),
+                    # Not anchored, page note
+                    (
+                        func.jsonb_array_length(Annotation.target_selectors) == 0,
+                        "page_note",
+                    ),
+                    # No annotation text, highlight
+                    (func.length(Annotation.text) == 0, "highlight"),
+                    # Anything else, an annotation
+                    else_="annotation",
+                ).label("type"),
+            )
+            .join(Annotation)
+            .join(User, User.id == AnnotationSlim.user_id)
             .join(Group, Group.id == AnnotationSlim.group_id)
-            .join(AnnotationMetadata)
+            .join(
+                AnnotationMetadata,
+                AnnotationSlim.id == AnnotationMetadata.annotation_id,
+            )
             .where(
                 # Visible annotations
                 AnnotationSlim.deleted == False,
                 AnnotationSlim.moderated == False,
                 AnnotationSlim.shared == True,
+                User.nipsa.is_(False),
                 # Limit search to the groups from the current authority
                 Group.authority == self._authorized_authority,
                 # From the groups we are interested
-                # Even if this is assignment centric an assigment
+                # Even if this is assignment centric an assignment
                 # might expand over multiple groups if using sections/groups
                 Group.authority_provided_id.in_(groups),
-                # And finally the assignment ID
+            )
+        )
+
+        if assignment_id:
+            query = query.where(
```

Comment:
Instead of having this condition here and an optional `assignment_id` param, you could leve only the common parts here, and move the `assignment_id` condition to `assignment_stats()`.

Reply:
Done